src/ztex/task.c

```diff
@@ -23,13 +23,50 @@
 
 
 
+static void task_result_list_init(struct task_result_list *list)
+{
+	list->count = 0;
+	list->result_list = NULL;
+	list->index = NULL;
+}
+
+
+static void task_result_list_add(struct task_result_list *list,
+		struct task_result *task_result)
+{
+	task_result->next = list->result_list;
+	list->result_list = task_result;
+	list->count++;
+	if (list->index) {
+		MEM_FREE(list->index);
+		list->index = NULL;
+	}
+}
+
+
+static void task_result_list_create_index(struct task_result_list *list)
+{
+	if (list->index) {
+		MEM_FREE(list->index);
+		list->index = NULL;
+	}
+	if (!list->count)
+		return;
+
+	list->index = mem_alloc(list->count * sizeof(struct task_result *));

```

Comment:
We already have dynamic memory allocation on every match (to be precise, on every application-level packet received from a board, be it a packet with a match or e.g. PROCESSING_DONE).
I've decided to allocate memory dynamically because it has to allow every candidate be a match (and it sometimes happen in tests) and the number of candidates is high (e.g. in descrypt 140 M/board, in md5crypt >0.5 M/board).

What could be an alternative approach? Some applications allocate a big chunk of memory and then perform management of allocated memory on its own. Because of application specifics such approach might be more efficient. In this application, memory management seem doesn't consume significant fraction of resources, I decided to use a ready-to-go allocator.

In this particular case, I think we can skip allocation and creation of index if the number of matches is low (e.g. 1-5).